scripts/image_downloader.py

```diff
@@ -0,0 +1,127 @@
+
+    def unpickle(file):
+        import pickle
+        with open(file, 'rb') as fo:
+            dict = pickle.load(fo, encoding='bytes')
+        return dict
+    
+    Loaded in this way, each of the batch files contains a dictionary with the following elements:
+    data -- a 10000x7500 numpy array of uint8s. Each row of the array stores a 80x80 colour image. The first 2500 entries contain the red channel values, the next 2500 the green, and the final 2500 the blue.
+    The image is stored in row-major order, so that the first 50 entries of the array are the red channel values of the first row of the image.
+    labels -- a list of 10000 numbers in the range 0-20. The number at index i indicates the label of the ith image in the array data.
+
+    Dependencies: Pillow, Pickle, numpy, requests
+"""
+
+from os import listdir
+import csv
+from enum import Enum
+import requests
+import numpy as np
+import pickle
+import io
+import PIL.Image as Image
+import PIL.ImageOps as ImageOps
+import threading
+from time import sleep
+class Tags(Enum):
+    NAME = 0
+    THUMBNAIL = 1
+    DATE = 2
+    IMG = 3
+    SUB_ID = 4
+    SUB = 5
+
+def download_file_images(file, data, labels, lock: threading.RLock):
+    global collected_samples
+    csv_filename = DATASET_DIR + file
+    with open(csv_filename, 'r', newline='') as csvfile:
+        reader = csv.reader(csvfile, delimiter=',')
+        #read each line
+        for index, row in enumerate(reader):
+            #take one in ten
+            if index % (100 / percentage2download) == 0:
+                date = row[Tags.DATE.value]
+                #check date
+                if not date or int(date.split("-")[0]) > date_accepted:
+                    thumb_url = row[Tags.THUMBNAIL.value]
+                    try:
+                        answer = requests.get(thumb_url, headers=headers) #download thumbnail
+                        if answer.status_code <= 400:
+                            image = answer.content
+                            #resize image
+                            image = Image.open(io.BytesIO(image))
+                            # image.save(thumb_url.split("/")[-1]) #was used to make sure the images are stored correctly
+                            image_size = (max(image.size), max(image.size))
+                            image = ImageOps.pad(image, image_size, color="white")
+                            image = np.array(image)
+                            #store resized image
+                            with lock:
+                                data.append(image)
+                                labels.append(categories[GET_CATEGORY(file)])
+                                collected_samples += 1
+
+                                #check if collected enough samples
+                                if collected_samples % samples4file == 0:
+                                    with open(FINAL_DATASET_PATH(), "w") as data_file:
```

Comment:
One issue that I decided to ignore because of the scope of this script is that at the end the data list is not emptied (though this is not good if we want all the dataset, if we only want a portion we are sure we'll have all the files with exactly 10000 samples)

Reply:
Ok!

Reply:
Do we want to solve the last issue I pointed out?

Reply:
I am fine living with this issue for now! 😊